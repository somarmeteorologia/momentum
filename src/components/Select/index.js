import React, { memo, useState, useContext } from 'react'
import styled, { css, ThemeContext } from 'styled-components'
import PropTypes from 'prop-types'
import v4 from 'uuid/v4'
import { ifProp, prop } from 'styled-tools'
import posed from 'react-pose'

import { Icon } from '@components/Icon'
import { Closeable } from '@components/Closeable'

const defaultValue = {
  label: 'Selecionar'
}

const Container = styled.div`
  width: ${ifProp('full', '100%', '240px')};
  display: flex;
  flex-direction: column;
  position: relative;
`

const Label = styled.span`
  margin-bottom: 8px;
  font-family: ${prop('theme.font.family.inter')};
  font-size: ${prop('theme.font.size.twelve')};
  font-weight: ${prop('theme.font.weight.bold')};
  color: ${ifProp(
    'disabled',
    prop('theme.field.text.disabled'),
    prop('theme.field.text.primary')
  )};
  cursor: default;

  ${ifProp(
    'required',
    css`
      &::after {
        content: '*';
        color: ${ifProp(
    'disabled',
    prop('theme.field.text.disabled'),
    prop('theme.field.text.danger')
  )};
      }
    `
  )}
`

const Selectable = styled.div`
  width: 100%;
  height: 40px;
  padding: 0 16px;
  font-size: ${prop('theme.font.size.fifteen')};
  border-radius: ${prop('theme.border.radius.four')};
  border: ${ifProp(
    'active',
    prop('theme.field.border.secondary'),
    prop('theme.field.border.primary')
  )};
  background-color: ${prop('theme.field.bg.primary')};
  color: ${prop('theme.field.text.quaternary')};
  outline: 0;
  appearance: none;
  cursor: ${ifProp('disabled', 'not-allowed', 'pointer')};
  position: relative;
  user-select: none;
  display: flex;
  justify-content: space-between;
  align-items: center;

  &:hover {
    border: ${ifProp(
    'disabled',
    prop('theme.field.border.primary'),
    prop('theme.field.border.secondary')
  )};
  }

  ${ifProp(
    'icon',
    css`
      padding-left: 36px;
    `
  )}
`

const Options = posed(styled.ul`
  width: 100%;
  top: calc(100% + 4px);
  position: absolute;
  border-radius: ${prop('theme.border.radius.four')};
  background-color: ${prop('theme.field.bg.secondary')};
  list-style: none;
  padding: 1px;
  border: ${prop('theme.field.border.tertiary')};
  box-shadow: ${prop('theme.field.shadow.primary')};
`)({
  visible: { applyAtStart: { display: 'block' }, opacity: 1 },
  hidden: { applyAtEnd: { display: 'none' }, opacity: 0 }
})

const Option = styled.li`
  display: flex;
  align-items: center;
  width: 100%;
  height: 40px;
  border-radius: ${prop('theme.border.radius.four')};
  padding-left: 14px;
  padding-right: 14px;
  font-size: ${prop('theme.font.size.fourteen')};
  font-weight: ${ifProp(
    'selected',
    prop('theme.font.weight.bold'),
    prop('theme.font.weight.regular')
  )};
  color: ${ifProp(
    'selected',
    prop('theme.field.text.active'),
    prop('theme.field.text.tertiary')
  )};
  cursor: pointer;

  &:hover {
    color: ${prop('theme.field.text.quaternary')};
    background-color: ${prop('theme.field.bg.hover')};
  }
`

const Content = styled.div`
  display: flex;
  align-items: center;
  left: 16px;
  position: absolute;
  height: 40px;
  line-height: 40px;
`

const Title = styled.span`
  font-size: ${prop('theme.font.size.fourteen')};
  font-weight: ${prop('theme.font.weight.regular')};
  color: ${ifProp(
    'hasSelected',
    prop('theme.field.text.selected'),
    prop('theme.field.text.secondary')
  )};

  ${ifProp(
    'disabled',
    css`
      color: ${prop('theme.field.text.disabled')};
    `
  )};
`

export const Select = memo(
  ({
    options,
    onChange,
    icon,
    defaultValue,
    label,
    disabled,
    required,
    full
  }) => {
    const { field } = useContext(ThemeContext)
    const [show, setShow] = useState(false)
    const [selected, setSelected] = useState(defaultValue)
    const [focus, setFocus] = useState(focus)

    const toggleShow = () => setShow(!show)
    const toCloseOutside = () => {
      setShow(false)
      setFocus(false)
    }

    const whenSelected = ({ text, value }) => () => {
      toggleShow()
      setSelected(text)
      onChange(value)
    }

    const whenFocus = () => {
      setFocus(!focus)
    }

    const hasSelected = () => Select.defaultValue.label !== selected

    return (
      <Container full={full}>
        {label && (
          <Label required={required} disabled={disabled} onClick={whenFocus}>
            {label}
          </Label>
        )}

        <Closeable whenClose={toCloseOutside}>
          <Selectable
            icon={icon}
            active={show || focus}
            onClick={!disabled ? toggleShow : undefined}
            disabled={disabled}
          >
            {icon && (
              <Content>
                {icon({ width: 14, height: 14, color: field.icon.primary })}
              </Content>
            )}
            <Title
              hasSelected={Select.defaultValue.label !== selected}
              disabled={disabled}
            >
              {selected}
            </Title>

            <Icon
              name={Icon.name.down}
              width={6}
              color={hasSelected() ? field.arrow.selected : field.arrow.primary}
            />
          </Selectable>

          <Options pose={show ? 'visible' : 'hidden'}>
            {options.map(({ value, text }) => (
              <Option
                key={v4()}
                selected={text === selected}
                value={value}
                onClick={whenSelected({ value, text })}
              >
                {text}
              </Option>
            ))}
          </Options>
        </Closeable>
      </Container>
    )
  }
)

Select.defaultProps = {
  defaultValue: defaultValue.label,
  onChange: () => {}
}

Select.propTypes = {
  defaultValue: PropTypes.string,
  icon: PropTypes.func,
  options: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired
}

Select.defaultValue = defaultValue
