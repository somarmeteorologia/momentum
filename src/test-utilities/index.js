import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import { Theme } from '@components/Theme'

export const renderWithTheme = (ui, theme = Theme.light) => {
  return {
    ...render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>)
  }
}
