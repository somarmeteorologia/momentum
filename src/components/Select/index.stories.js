import React from 'react'
import styled from 'styled-components'
import { storiesOf } from '@storybook/react'

import { GROUPS } from '@helpers/hierarchySeparators'
import { Container } from '@helpers/components/Container'

import { Reset } from '@components/Reset'
import { Icon } from '@components/Icon'
import { Select } from '@components/Select'

const Containerable = styled(Container)`
  flex-wrap: wrap;
`

const Wrapper = styled.div`
  display: flex;
  align-items: flex-end;
`

storiesOf(`${GROUPS.COMPONENTS}|Select`, module)
  .addDecorator(story => (
    <Containerable>
      <Reset />
      {story()}
    </Containerable>
  ))
  .add('Default', () => (
    <>
      <Select
        raw={true}
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
      />
    </>
  ))
  .add('With icon', () => (
    <>
      <Select
        icon={({ color, width, height }) => (
          <Icon
            name={Icon.name.checked}
            color={color}
            width={width}
            height={height}
          />
        )}
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
      />
    </>
  ))
  .add('With label', () => (
    <>
      <Select
        label="Select label"
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
      />

      <Select
        label="Select label"
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
        required
      />
    </>
  ))
  .add('Full', () => (
    <>
      <Select
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
        full
      />
    </>
  ))
  .add('Disabled', () => (
    <Wrapper>
      <Select
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
        disabled
      />
      <Select
        label="Select label"
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
        disabled
      />

      <Select
        label="Select label"
        options={[
          { value: 1, text: 'Banana' },
          { value: 2, text: 'Apple' },
          { value: 3, text: 'Orange' }
        ]}
        required
        disabled
      />
    </Wrapper>
  ))
