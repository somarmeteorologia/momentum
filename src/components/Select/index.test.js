import React from 'react'

import { renderWithTheme } from '../../test-utilities'

import { Select } from '@components/Select'

test('Not expect visual regression', () => {
  const { container } = renderWithTheme(
    <Select
      options={[
        { value: 1, text: 'Banana' },
        { value: 2, text: 'Apple' }
      ]}
    />
  )

  expect(container.firstChild).toMatchSnapshot()
})
