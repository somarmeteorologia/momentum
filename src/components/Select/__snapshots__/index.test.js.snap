// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Not expect visual regression 1`] = `
.c5 {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.c4 {
  -webkit-transform: scaleY(-1);
  -ms-transform: scaleY(-1);
  transform: scaleY(-1);
}

.c3 {
  display: inline-block;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 0;
}

.c0 {
  width: 240px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
}

.c1 {
  width: 100%;
  height: 40px;
  padding: 0 16px;
  border-radius: 0.25rem;
  border: 1px solid #b1bbe5;
  background-color: #f1f3fA;
  color: #fff;
  outline: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:hover {
  border: 1px solid #6479cb;
}

.c6 {
  width: 100%;
  top: calc(100% + 4px);
  position: absolute;
  border-radius: 0.25rem;
  background-color: #fff;
  list-style: none;
  padding: 1px;
  border: 1px solid #f1f3fA;
  box-shadow: 0 2px 4px rgba(50,71,151,0.15);
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 40px;
  border-radius: 0.25rem;
  padding-left: 14px;
  padding-right: 14px;
  font-size: 0.875rem;
  font-weight: 500;
  color: #19244d;
  cursor: pointer;
}

.c7:hover {
  color: #fff;
  background-color: #133BEC;
}

.c2 {
  font-size: 0.875rem;
  font-weight: 500;
  color: #6479cb;
}

<div
  class="c0"
>
  <div>
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        Selecionar
      </span>
      <div
        class="sc-AxjAm c3"
        height="20"
        width="6"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <svg
              fill="none"
              height="20"
              viewBox="0 0 9 15"
              width="6"
            >
              <path
                d="M1.55469 14L7.33247 7.5L1.55469 1"
                stroke="#6479cb"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-miterlimit="10"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <input
      type="hidden"
      value="Selecionar"
    />
    <div
      class="c6"
    >
      <div
        class="c7"
        value="1"
      >
        Banana
      </div>
      <div
        class="c7"
        value="2"
      >
        Apple
      </div>
    </div>
  </div>
</div>
`;
